### PR TITLE
Verbs provider. EP_RDM/FI_TAGGED. fi_tinject fix + ack handling opt

### PR DESCRIPTION
**Pull request by ddurnov**
Squashed #1512 + signature.

Signed-off-by: Dmitry Durnov <dmitry.durnov@intel.com>

### CODE REVIEW

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 1]
```diff
@@ -366,7 +366,113 @@ int fi_ibv_rdm_tagged_repost_receives(struct fi_ibv_rdm_tagged_conn *conn,
...
+				    i * ep->buff_len, BUF_STATUS_FREE);
+			}
+		}
+
+		fi_ibv_rdm_tagged_set_buffer_status(conn->sbuf_head,
+						    BUF_STATUS_BUSY);
+
+		sbuf = conn->sbuf_head;
+		fi_ibv_rdm_tagged_push_sbuff_head(conn, ep);
+	}
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: indentation is off

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 2]
```diff
@@ -366,7 +366,113 @@ int fi_ibv_rdm_tagged_repost_receives(struct fi_ibv_rdm_tagged_conn *conn,
...
+fi_ibv_rdm_tagged_prepare_send_resources(struct fi_ibv_rdm_tagged_conn *conn,
+					 struct fi_ibv_rdm_ep *ep)
+{
+	if (conn->state != FI_VERBS_CONN_ESTABLISHED) {
+		pthread_mutex_lock(&ep->cm_lock);
+		if (conn->state == FI_VERBS_CONN_ALLOCATED) {
+			fi_ibv_rdm_start_connection(ep, conn);
+		}
+		pthread_mutex_unlock(&ep->cm_lock);
+		usleep(1000);
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: ?

File: prov/verbs/src/ep_rdm/verbs_utils.h [hunk 3]
```diff
@@ -45,6 +45,20 @@
...
 #include "../fi_verbs.h"
 
+#if defined(__ICC) || defined(__INTEL_COMPILER) || \
+ defined(__GNUC__) || defined(__GNUG__)
+#include "xmmintrin.h"
+#endif /* ICC || GCC */
+
+#if defined(__ICC) || defined(__INTEL_COMPILER) || \
+ defined(__GNUC__) || defined(__GNUG__)
+#define FI_IBV_PREFETCH_ADDR(_addr) {                    \
```
Comment (shefty) [inline prov/verbs/src/ep_rdm/verbs_utils.h]: Move the include from above here to avoid listing the same set of defines twice in a row.

**Comment by shefty** [comment]
For the verbs RDM support, is there anyone specific that you want to ack the changes before they are merged?

File: prov/verbs/src/ep_rdm/verbs_utils.h (hunk 3 above)
Comment (ddurnov) [inline prov/verbs/src/ep_rdm/verbs_utils.h]: OK

File: prov/verbs/src/ep_rdm/verbs_rdm.h [hunk 4]
```diff
@@ -366,7 +366,113 @@ int fi_ibv_rdm_tagged_repost_receives(struct fi_ibv_rdm_tagged_conn *conn,
...
+	int i = 0;
+	void *sbuf = NULL;
+
+	if (fi_ibv_rdm_tagged_get_buffer_status(conn->sbuf_head) ==
+	    BUF_STATUS_FREE) {
+
+		/* We have made whole circle. Reset buffer states */
+		if (conn->sbuf_head ==
+		    (conn->sbuf_mem_reg +
+		     FI_IBV_RDM_TAGGED_BUFF_SERVICE_DATA_SIZE)) {
```
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_rdm.h]: Minor: would be nice to have getter for the first sbuf, or even generalized one by sequence number like rbuf has if it does not affect performance

File: prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c [hunk 5]
```diff
@@ -332,6 +299,10 @@ static ssize_t fi_ibv_rdm_tagged_senddatato(struct fid_ep *fid, const void *buf,
 			.rndv_threshold = ep->rndv_threshold,
 			.imm = (uint32_t) data
 		};
+
+		request->state.eager = 0; /* Initial state */
+		request->state.rndv  = 0;
```
Comment (evgeny-leksikov) [inline prov/verbs/src/ep_rdm/verbs_tagged_ep_rdm.c]: MInor: initialize by correspond enum values

**Comment by evgeny-leksikov** [comment]
generally OK for me, just a couple of minor comments.

**Comment by shefty** [comment]
Please update based on Evgeny's comments.  I would like to merge and create 1.2rc2 on 12/8.

**Comment by ddurnov** [comment]
Done

**Pull request merged**